Pat.Subscriber.Tools/AuthTokenProvider.cs

```diff
@@ -0,0 +1,110 @@
+﻿using System;
+using System.IdentityModel.Tokens.Jwt;
+using System.Linq;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Threading.Tasks;
+using Microsoft.IdentityModel.Clients.ActiveDirectory;
+
+namespace Pat.Subscriber.Tools
+{
+    public class AuthTokenProvider
+    {
+        private string _tokenValue = string.Empty;
+        private DateTime _tokenExpiresAtUtc = DateTime.MinValue;
+
+        private string _clientId;
+        private string _clientSecret;
+        private string _tenantId;
+        private readonly FileTokenCache _tokenCache;
+
+        private const string PowerShellClientId = "1950a258-227b-4e31-a9cf-717495945fc2";
+        private const string AzureManagementResource = "https://management.azure.com/";
+
+        public AuthTokenProvider()
+        {
+            _tokenCache = new FileTokenCache();
+        }
+
+
+        public string AuthUser
+        {
+            get
+            {
+                var handler = new JwtSecurityTokenHandler();
+                var token = (JwtSecurityToken)handler.ReadToken(_tokenValue);
+                var name = token.Claims.SingleOrDefault(c => c.Type == "name");
+                var upn = token.Claims.SingleOrDefault(c => c.Type == "upn");
+                var appId = token.Claims.SingleOrDefault(c => c.Type == "appid");
+                return string.IsNullOrEmpty($"{appId.Value}") 
+                    ? $"{name.Value} <{upn.Value}>"
+                    : $"{appId.Value}";
+            }
+        }
+
+        public void SetServicePrincipal(string configCommandClientId, string configCommandClientSecret, string tenantId)
+        {
+            _clientId = configCommandClientId;
+            _clientSecret = configCommandClientSecret;
+            _tenantId = tenantId;
+        }
+
+        public async Task EnsureAuthTokenValid(HttpClient client)
+        {
+            // Check to see if the token has expired before requesting one.
+            // We will go ahead and request a new one if we are within 2 minutes of the token expiring.
+            if (_tokenExpiresAtUtc < DateTime.UtcNow.AddMinutes(-2))
+            {
```

Comment:
Pedantisticicity warning....  arrowhead - maybe invert the "if" and return early instead of indenting the rest of the method?

Reply:
I did think about that, but I wanted to limit the changes I introduced. I just wanted to reduce the awful in `AzureHttpClient`. I'll do that in a future pass #probablynot